File: apps/api/app/schemas/concept.py
from pydantic import BaseModel, ConfigDict

from app.models.enums import ConceptEdgeRelationshipType


class ConceptSummaryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    slug: str
    name: str
    topic_id: int
    topic_slug: str


class ConceptRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    slug: str
    name: str
    topic_id: int
    topic_slug: str
    definition: str | None
    formula: str | None
    intuition: str | None
    common_mistakes: str | None
    interview_tips: str | None
    prerequisites: str | None


class ConceptNeighborRead(BaseModel):
    slug: str
    name: str
    relationship_type: ConceptEdgeRelationshipType
    direction: str


class ConceptDetailRead(ConceptRead):
    neighbors: list[ConceptNeighborRead]

File: apps/web/lib/types/concept.ts
export type ConceptSummary = {
  id: number;
  slug: string;
  name: string;
  topic_id: number;
  topic_slug: string;
};

export type ConceptRelationshipType = "requires" | "related_to" | "used_in";

export type ConceptNeighbor = {
  slug: string;
  name: string;
  relationship_type: ConceptRelationshipType;
  direction: "incoming" | "outgoing";
};

export type ConceptDetail = ConceptSummary & {
  definition: string | null;
  formula: string | null;
  intuition: string | null;
  common_mistakes: string | null;
  interview_tips: string | null;
  prerequisites: string | null;
  neighbors: ConceptNeighbor[];
};
